lib/client.js

```diff
@@ -433,10 +435,18 @@ MqttClient.prototype.publish = function (topic, message, opts, callback) {
     case 2:
       // Add to callbacks
       this.outgoing[packet.messageId] = callback || nop
-      this._sendPacket(packet, undefined, opts.cbStorePut)
+      if (this.storeProcessing) {
+        this._storePacket(packet, undefined, opts.cbStorePut)
+      } else {
+        this._sendPacket(packet, undefined, opts.cbStorePut)
+      }
       break
     default:
-      this._sendPacket(packet, callback, opts.cbStorePut)
+      if (this.storeProcessing) {
+        this._storePacket(packet, callback, opts.cbStorePut)
+      } else {
+        this._sendPacket(packet, callback, opts.cbStorePut)
+      }
```

Comment:
When will these packets be published if we just store them?

Reply:
The stored packet will be published at https://github.com/mqttjs/MQTT.js/blob/master/lib/client.js#L1334. `this._storePacket()` is called when the flag `storeProcessing` is true. That means publish stored packet process started but not finished yet. So the stored packet will processed before [`on('end'...)`](https://github.com/mqttjs/MQTT.js/blob/master/lib/client.js#L1294)  will happen.